adcc/OneParticleOperator.py

```diff
@@ -52,41 +57,183 @@ def __init__(self, spaces, is_symmetric=True):
             assert isinstance(spaces.reference_state, libadcc.ReferenceState)
             self.reference_state = spaces.reference_state
 
-    @classmethod
-    def from_cpp(cls, cpp_operator):
-        assert cpp_operator.cartesian_transform == "1"
-        ret = cls(cpp_operator.mospaces, cpp_operator.is_symmetric)
-        for b in cpp_operator.blocks_nonzero:
-            ret.set_block(b, cpp_operator.block(b))
+        occs = sorted(self.mospaces.subspaces_occupied, reverse=True)
+        virts = sorted(self.mospaces.subspaces_virtual, reverse=True)
+        self.orbital_subspaces = occs + virts
+        if self.is_symmetric:
+            combs = list(
+                combinations_with_replacement(self.orbital_subspaces, r=2)
+            )
+        else:
+            combs = list(product(self.orbital_subspaces, repeat=2))
+        self.blocks = ["".join(com) for com in combs]
+        self._tensors = {}
+
+    @property
+    def shape(self):
+        """
+        Returns the shape tuple of the OneParticleOperator
+        """
+        size = sum(self.mospaces.n_orbs(ss) for ss in self.orbital_subspaces)
+        return (size, size)
+
+    @property
+    def size(self):
+        """
+        Returns the number of elements of the OneParticleOperator
+        """
+        return np.prod(self.shape)
+
+    @property
+    def blocks_nonzero(self):
+        """
+        Returns a list of the non-zero block labels
+        """
+        return [b for b in self.blocks if b in self._tensors]
+
+    def is_zero_block(self, block):
+        """
+        Checks if block is explicitly marked as zero block.
+        Returns False if the block does not exist.
+        """
+        if block not in self.blocks:
+            return False
+        return block not in self.blocks_nonzero
+
+    def block(self, block):
+        """
+        Returns tensor of the given block
+        """
+        if block not in self.blocks_nonzero:
+            raise KeyError("The block function does not support "
+                           "access to zero-blocks. Available non-zero "
+                           f"blocks are: {self.blocks_nonzero}.")
+        return self._tensors[block]
+
+    def __getitem__(self, block):
+        if block not in self.blocks:
+            raise KeyError(f"Invalid block {block} requested. "
+                           f"Available blocks are: {self.blocks}.")
+        if block not in self._tensors:
+            sym = libadcc.make_symmetry_operator(
+                self.mospaces, block, self.is_symmetric, "1"
+            )
+            self._tensors[block] = Tensor(sym)
+        return self._tensors[block]
+
+    def __setitem__(self, block, tensor):
+        self.set_block(block, tensor)
+
+    def set_block(self, block, tensor):
```

Comment:
Do we still need the `set_block` function at all? Maybe we should make it deprecated and remove it soon?

Reply:
Yes, I completely agree. Since everything is on the py-side now, no reason to keep it.